### PR TITLE
Selectively enable different frontends

### LINKED ISSUE #1943: ✨[Feature] Remove C++ Dependency in Dynamo
## Context + Feature Description
Currently, `python setup.py install --fx-only` does not apply for Dynamo due to the presence of certain operators such as `torch_tensorrt._Input`. Input, device, and other data structures which depend on the C++ portion of the API can be eliminated by making Python-only versions of these data structures in the `torch_tensorrt.dynamo` module directory. Then, Dynamo can be imported directly using only Python installs, which are much faster and easier than the C++ install.

### LINKED ISSUE #2379: Decoupling the torchscript and dynamo subpackages in python

### PR DESCRIPTION
**Pull request by narendasan**
# Description

Allows users to configure which frontends/features they want to include in the `torch-tensorrt` builds.

Features can be enabled as so: 
```py
NO_TORCHSCRIPT=1 pip install -e . ... # Includes the C++ runtime but no Torchscript frontend
PYTHON_ONLY=1 pip install -e . ... # No C++ dependencies at all, a pure python package 
```

A builds feature set can be accessed via the following struct 

```py
import torch_tensorrt

torch_tensorrt.ENABLED_FEATURES
```

where `ENABLED_FEATURES` is a `namedtuple` `FeatureSet`: 
```py
namedtuple('FeatureSet', [
    "torchscript_frontend",
    "torch_tensorrt_runtime",
    "dynamo_frontend",
    "fx_frontend"
])
```

In order to support optional features, a number of core types have been abstracted: 

- `torch_tensorrt.Device` has no direct dependencies on the torchscript core and can be translated to `torch_tensorrt.ts.Device` to access those features
- Similarly `torch_tensorrt.Input` behaves the same way
- enums for `dtype`,`DeviceType`, `memory_format` have been defined and can translate from `numpy`, `tensorrt`, `torch` and `torch_tensorrt._C` (assuming `torch_tensorrt.ENABLED_FEATURES.torchscript_frontend` is `True`)

Translating between different library enums now can take the form
```py
import torch
import tensorrt as trt
import torch_tensorrt

# Convert, throw error if no conversion exists
trt_dtype = torch_tensorrt.dtype._from(torch.float16).to(trt.DataType) # returns trt.DataType.HALF

#Try to convert, do not throw an error if there is no conversion, just return None
trt_dtype = torch_tensorrt.dtype.try_from(torch.float16).to(trt.DataType) # returns trt.DataType.HALF

# Alternatively, use a fallback type
trt_dtype = torch_tensorrt.dtype.unknown.to(trt.DataType, use_default=True) #returns trt.DataType.FLOAT
```

Fixes #1943
Fixes #2379 
 

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/__init__.py [hunk 1]
```diff
@@ -80,26 +80,41 @@ def _find_lib(name: str, paths: List[str]) -> str:
...
 def _register_with_torch() -> None:
     trtorch_dir = os.path.dirname(__file__)
-    torch.ops.load_library(trtorch_dir + "/lib/libtorchtrt.so")
+    if os.path.isfile(trtorch_dir + "/lib/libtorchtrt.so"):
+        assert ENABLED_FEATURES.torchscript_frontend == True
+        assert ENABLED_FEATURES.torch_tensorrt_runtime == True
+        torch.ops.load_library(trtorch_dir + "/lib/libtorchtrt.so")
+    elif os.path.isfile(trtorch_dir + "/lib/libtorchtrt_runtime.so"):
+        assert ENABLED_FEATURES.torch_tensorrt_runtime == True
+        torch.ops.load_library(trtorch_dir + "/lib/libtorchtrt_runtime.so")
```
Comment (gs-olive) [inline py/torch_tensorrt/__init__.py:100]: Can omit `== True` from assertions

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 2]
```diff
@@ -198,25 +198,18 @@ def compile(
...
-        torch.float16 in enabled_precisions
-        or torch_tensorrt.dtype.half in enabled_precisions
-    ):
-        precision = torch.float16
-    elif (
-        torch.float32 in enabled_precisions
-        or torch_tensorrt.dtype.float in enabled_precisions
-    ):
-        precision = torch.float32
+    if dtype.float16 in enabled_precisions or dtype.half in enabled_precisions:
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_compiler.py:201]: `dtype.float16` and `dtype.half` seem to point to the same `enum` object. Should this be:

```python
if dtype.float16 in enabled_precisions or torch.float16 in enabled_precisions:
```

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 3]
```diff
@@ -198,25 +198,18 @@ def compile(
...
+        precision = dtype.float16
+    elif dtype.float32 in enabled_precisions or dtype.float in enabled_precisions:
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_compiler.py:203]: Same as above comment

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 4]
```diff
@@ -569,24 +564,15 @@ def convert_module_to_trt_engine(
...
+        precision = dtype.float32
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_compiler.py:575]: Same as above

File: py/torch_tensorrt/dynamo/_settings.py [hunk 5]
```diff
@@ -74,7 +73,7 @@ class CompilationSettings:
         output_format (str): Output format of the result of TRT compilation. Options include "exported_program" (or) "ep" | "torchscript" (or) "ts" | "graph_module" (or) "fx". Default is "exported_program"
     """
 
-    precision: torch.dtype = PRECISION
+    precision: dtype = field(default_factory=lambda: PRECISION)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_settings.py:76]: What is the reasoning for using the `field` here?

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 6]
```diff
@@ -201,12 +203,20 @@ def run(
...
             self.compilation_settings.num_avg_timing_iters
         )
 
         if self.compilation_settings.device.device_type == trt.DeviceType.DLA:
+            device_info = torch.cuda.get_device_properties(
+                self.compilation_settings.device.gpu_id
+            )
+            assert (device_info.major == 8 and device_info.minor == 7) or (
+                device_info.major == 7 and device_info.minor == 2
+            )
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:219]: Just to verify - should these be `>=` ranges, or is it a fixed set?

**Review by gs-olive — COMMENTED** [review]
Overall looks great; overhaul of dtype system and frontend selection at install is very helpful. I just need to verify on local installs as well. Left a few small comments

File: py/torch_tensorrt/dynamo/_settings.py (hunk 5 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/_settings.py:76]: Its a WAR dealing with enums not being immutable

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py (hunk 6 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:219]: Its a fixed set these are 2 specific SM versions that key to AGX Orin and AGX Xavier

File: py/torch_tensorrt/_compile.py [hunk 7]
```diff
@@ -261,7 +281,7 @@ def convert_method_to_trt_engine(
     method_name: str = "forward",
     inputs: Optional[Sequence[Input | torch.Tensor]] = None,
     ir: str = "default",
-    enabled_precisions: Optional[Set[torch.dtype | dtype]] = None,
+    enabled_precisions: Optional[Set[torch.dtype]] = None,
```
Comment (narendasan) [inline py/torch_tensorrt/_compile.py:284]: Correct type annotation

File: py/torch_tensorrt/_compile.py [hunk 8]
```diff
@@ -299,21 +319,22 @@ def convert_method_to_trt_engine(
...
             ts_mod,
             inputs=inputs,
             method_name=method_name,
             enabled_precisions=enabled_precisions_set,
             **kwargs,
         )
+        return serialized_engine
     elif target_ir == _IRType.fx:
         raise RuntimeError(
             "convert_method_to_trt_engine call is not supported for ir=fx"
```
Comment (narendasan) [inline py/torch_tensorrt/_compile.py:340]: adjust name of dynamo converte_module

File: py/torch_tensorrt/_compile.py [hunk 9]
```diff
@@ -167,13 +187,13 @@ def compile(
     Returns:
         torch.nn.Module: Compiled Module, when run it will execute via TensorRT
     """
-    input_list = inputs if inputs is not None else []
-    enabled_precisions_set = (
-        enabled_precisions if enabled_precisions is not None else {torch.float}
+    input_list = inputs if inputs else []
```
Comment (gs-olive) [inline py/torch_tensorrt/_compile.py:190]: Consider switching to `if inputs is not None`

File: py/torch_tensorrt/_enums.py [hunk 10]
```diff
@@ -1,3 +1,707 @@
...
+        self,
+        t: Union[
+            Type[torch.memory_format], Type[trt.TensorFormat], Type[memory_format]
+        ],
+    ) -> Optional[Union[torch.memory_format, trt.TensorFormat, memory_format]]:
+        try:
+            casted_format = self.to(t)
+            return casted_format
+        except (ValueError, TypeError) as e:
+            logging.debug(e)
```
[the remark this answers is not in the document]
Comment (gs-olive) [reply py/torch_tensorrt/_enums.py:495]: Similarly for other `trt_*` functions

File: py/torch_tensorrt/dynamo/conversion/converter_utils.py [hunk 11]
```diff
@@ -268,7 +265,9 @@ def create_constant(
     Returns:
         A TensorRT ITensor that represents the given value.
     """
-    numpy_value = to_numpy(value, dtype)
+    numpy_value = to_numpy(
+        value, _enums.dtype._from(dtype).to(np.dtype) if dtype else None
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/converter_utils.py:269]: Consider switching to `if dtype is not None`

**Comment by gs-olive** [comment]
Additionally, tested on Windows E2E models and appears to cleanly dispatch to the correct runtime without needing to modify the `use_python_runtime` flag

**Comment by gs-olive** [comment]
A few examples, such as the one below, use `precision` still and would fail with some of the changes.

https://github.com/pytorch/TensorRT/blob/c5b8909f47485df934041e8fd40e1f75f9d77671/examples/dynamo/torch_compile_stable_diffusion.py#L42

File: py/torch_tensorrt/_enums.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (narendasan) [reply py/torch_tensorrt/_enums.py]: I think root logger is fine for this because theres only types no functional code

File: toolchains/legacy/pyproject.toml [hunk 12]
```diff
@@ -64,7 +64,7 @@ include-package-data = false
```
Comment (peri044) [inline toolchains/legacy/pyproject.toml:64]: Where would this file be used ?

**Review by peri044 — APPROVED** [review]
LGTM. Added minor comments

**Comment by peri044** [comment]
Will docs be updated with the instructions as a different PR ?
